src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AwsHttpCredentialsAdapter.java

```diff
@@ -0,0 +1,542 @@
+    }
+  }
+
+  // Utility methods for supporting AWS signatures
+  //------------------------------------------------------------------------------------------
+  private Stream<Map.Entry<String, String>> processHeaders(final HttpRequest request, final String amzDate) {
+    final Map<String, List<String>> headers = new HashMap<>();
+    request.headers()
+        .iteratorAsString()
+        .forEachRemaining(x -> headers
+            .computeIfAbsent(x.getKey().toLowerCase(), _x -> new ArrayList<>())
+            .add(x.getValue()));
+
+    Preconditions.checkNotNull(headers.get("host"));
+    if (useUnsignedPayloads) {
+      headers.putIfAbsent(AWS_CONTENT_SHA_HDR, UNSIGNED_PAYLOAD_SIG);
+    }
+    headers.putIfAbsent(AWS_DATE_HDR, ImmutableList.of(amzDate));
+
+    return headers
+        .entrySet()
+        .stream()
+        .filter(x -> x.getKey() != null)
+        .filter(x -> x.getKey().contains(AWS_HDR_PREFIX) || headersToInclude.contains(x.getKey()))
+        .map(this::processHeader)
+        .sorted((x, y) ->
+          ComparisonChain.start()
+              .compare(x.getKey(), y.getKey())
+              .compare(x.getValue(), y.getValue())
+            .result()
+        );
+  }
+
+  private Map.Entry<String, String> processHeader(final Map.Entry<String, List<String>> header) {
+    final String headerKey = header.getKey().toLowerCase();
+    final String headerValue = processHeaderValues(header.getValue());
+    return Maps.immutableEntry(headerKey, headerValue);
+  }
+
+  private String processHeaderValues(final List<String> values) {
+    return COMMA_JOINER.join(values.stream().map(this::trimSpaces).iterator());
+  }
+
+  private String trimSpaces(final String value) {
+    return CharMatcher.whitespace().trimAndCollapseFrom(value, ' ');
+  }
+
+  private byte[] sha256Hash(final String canonicalRequest) throws NoSuchAlgorithmException {
+    final MessageDigest digest = MessageDigest.getInstance("SHA-256");
+    return digest.digest(canonicalRequest.getBytes(Charsets.UTF_8));
+  }
+
+  private CharSequence hex(final byte[] bytes) {
+    return HashCode.fromBytes(bytes).toString();
+  }
+
+  byte[] hmacSHA256(final String key, final String toSign) throws IOException {
+    return hmacSHA256(key.getBytes(Charsets.UTF_8), toSign.getBytes(Charsets.UTF_8));
+  }
+
+  byte[] hmacSHA256(final byte[] key, final String toSign) throws IOException {
+    return hmacSHA256(key, toSign.getBytes(Charsets.UTF_8));
+  }
+
+  byte[] hmacSHA256(final byte[] key, final byte[] toSign) throws IOException {
+    try {
+      final Mac mac = Mac.getInstance(SIGNING_ALGO);
+      mac.init(new SecretKeySpec(key, SIGNING_ALGO));
+      return mac.doFinal(toSign);
+    } catch (NoSuchAlgorithmException e) {
+      throw new IOException("Could not initialise crypto algorithms", e);
+    } catch (InvalidKeyException e) {
+      throw new IOException("Provided secret key was invalid for crypto purposes", e);
+    }
+  }
+}
+
+enum AwsRegion {
```

Comment:
Where did you take these information from? Would there be a way to automatically generate this enum?

Reply:
Its from the API docs and boto right now, I dont know of an easy way to autogenerate this :(

Reply:
I found an enum in the SDK that *should* work for our use cases, I have converted to using this